src/internal/operators/windowWhen.ts

```diff
@@ -50,95 +51,76 @@ import { lift } from '../util/lift';
  * are Observables.
  * @name windowWhen
  */
-export function windowWhen<T>(closingSelector: () => Observable<any>): OperatorFunction<T, Observable<T>> {
-  return function windowWhenOperatorFunction(source: Observable<T>) {
-    return lift(source, new WindowOperator<T>(closingSelector));
-  };
-}
-
-class WindowOperator<T> implements Operator<T, Observable<T>> {
-  constructor(private closingSelector: () => Observable<any>) {
-  }
-
-  call(subscriber: Subscriber<Observable<T>>, source: any): any {
-    return source.subscribe(new WindowSubscriber(subscriber, this.closingSelector));
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class WindowSubscriber<T> extends ComplexOuterSubscriber<T, any> {
-  private window: Subject<T> | undefined;
-  private closingNotification: Subscription | undefined;
-
-  constructor(protected destination: Subscriber<Observable<T>>,
-              private closingSelector: () => Observable<any>) {
-    super(destination);
-    this.openWindow();
-  }
-
-  notifyNext(_outerValue: T, _innerValue: any,
-             _outerIndex: number,
-             innerSub: ComplexInnerSubscriber<T, any>): void {
-    this.openWindow(innerSub);
-  }
-
-  notifyError(error: any): void {
-    this._error(error);
-  }
-
-  notifyComplete(innerSub: ComplexInnerSubscriber<T, any>): void {
-    this.openWindow(innerSub);
-  }
+export function windowWhen<T>(closingSelector: () => ObservableInput<any>): OperatorFunction<T, Observable<T>> {
+  return (source: Observable<T>) =>
+    lift(source, function (this: Subscriber<Observable<T>>, source: Observable<T>) {
+      const subscriber = this;
+      let window: Subject<T>;
+      let closingSubscriber: Subscriber<any>;
 
-  protected _next(value: T): void {
-    this.window!.next(value);
-  }
+      /**
+       * When we get an error, we have to notify both the
+       * destiation subscriber and the window.
+       */
+      const handleError = (err: any) => {
+        window.error(err);
+        subscriber.error(err);
+      };
 
-  protected _error(err: any): void {
-    this.window!.error(err);
-    this.destination.error(err);
-    this.unsubscribeClosingNotification();
-  }
+      /**
+       * Called every time we need to open a window.
+       * Recursive, as it will start the closing notifier, which
+       * inevitably *should* call openWindow -- but may not if
+       * it is a "never" observable.
+       */
+      const openWindow = () => {
+        // We need to clean up our closing subscription,
+        // we only cared about the first next or complete notification.
+        closingSubscriber?.unsubscribe();
 
-  protected _complete(): void {
-    this.window!.complete();
-    this.destination.complete();
-    this.unsubscribeClosingNotification();
-  }
+        // Close our window before starting a new one.
+        window?.complete();
```

Comment:
issue: AFAICT, this optional chaining is unnecessary. I can't see that `window` is nulled out anyplace. Maybe it should be? And if it is, the type should reflect that, IMO.